### DISCUSSION TITLE
Using npm packages as extensions

### DISCUSSION THREAD

**Discussion by nickrum** [directus/directus#3056] [Feature Requests]
I had this idea about extensions being npm packages. I think using npm packages would make the extension system much more flexible and powerful. VS Code does this for their extensions and it works quite well.
Since "create-directus-project" already creates a npm package, extensions could simply be installed as dependencies to this package and then be loaded by the API.

Some advantages I can think of:
- It would make directus/next#255 trivial by enforcing a naming pattern like directus-extension-* for extensions, kind of like rollup does it with rollup-plugin-*.
- It would help with directus/directus#3044 since the dev environment could simply be installed as dev dependencies of the package.
- It would help with directus/directus#3050 because installing packages in contrast to copy-pasting folders would make it less likely that files of extensions consisting of multiple files get lost in the process.
- Extensions could be installed in every way a npm package can be installed, like a tarball, a git repo, a folder or a remote url.
- Extensions could easily be versioned and updated.
- Private npm repositories could be used to deliver extensions. A marketplace could also be realized as a private repository.

Some disadvantages I can think of:
- Extensions wouldn't be as simple as being a single file.
- Extensions couldn't easily be loaded from folders like it is currently done.

To mitigate the disadvantages for developers, a CLI like "create-directus-extension" could help with scaffolding the extension package. Extensions for the App already have to be developed as npm packages because of the required build step, so this would not make that much of a difference.

To mitigate the disadvantages for users, the directus CLI could provide a command that installs all extensions inside a provided folder as dependencies to the current directus project. This would also be useful for the docker image. Maybe custom hooks (and maybe even custom API endpoints) could still be loaded from a certain folder to make basic extensibility more approachable.

While this is a major change quite far into the development process, I think something like this would be worth it. What do you think?

**Comment by rijkvanzanten** [comment]
Yes absolutely, this is actually the main strategy I've got for "extension bundles" and the in-app marketplace 🙂  Glad to hear that we're not the only one considering this route. As for

> While this is a major change quite far into the development process, I think something like this would be worth it. What do you think?

I don't think it has to be a major change. I see the npm distributed packages as an addition to 'local extensions', rather than a replacement. I'd still like you to be able to simply throw in a file to get a custom hook running instead of having to go through the whole flow of setting up a new project, building it, publishing it to npm, installing it in your project etc etc

**Reply by nickrum** [reply]
That's very exciting to hear! Absolutely not, I have worked with a npm-package-based plugin system before and I found it to be really convenient to re-use the infrastructure that is already there.

It would of course be possible to support both methods but packaged extensions might benefit from additional data that could be stored inside _package.json_.
I absolutely agree with you that custom hooks should be as simple to create and install as possible. How about the other types of extensions?
I think npm packages can also be installed locally without building and publishing but I see your point 🙂

**Reply by rijkvanzanten** [reply]
As of right now, I think the plan is to leave extensions "as-is", add support for "extension bundles" (where you can have any extension combined into a single distribution), and then add support for auto-reading all packages that start with a prefix like `directus-extension-*` from `node_modules` and consider those "bundles". That way, we have can reuse the same logic between "local extensions", "extension bundles", and "npm extensions" 🤔

**Reply by nickrum** [reply]
That sounds like a good plan. Especially regarding the reuse of the same logic between all "types" of extensions.

My motivation for starting a discussion on using a slightly different system was to be able to load App extensions on demand and reduce the initial loading time of the whole App.

**Reply by rijkvanzanten** [reply]
> My motivation for starting a discussion on using a slightly different system was to be able to load App extensions on demand and reduce the initial loading time of the whole App.

Care to elaborate on this? Are you talking about asynchronously downloading the app-extensions so the app loads faster? (It currently does that 😁)

**Reply by nickrum** [reply]
Sure. It does indeed, but all extensions have to be downloaded for the app to load, right? 😉 I was thinking of something like _async components + code splitting_ but for extensions.

**Reply by rijkvanzanten** [reply]
I built it that way in v7 actually, where interfaces / displays where only loaded the first time they were used to speed up the rest of the app load. However, that resulted in a very janky user experience with things loading randomly and jumping in 🤔 I'm sure we can find a good middleground though. Generally speaking, having a longer load up front and then a smoother experience when using the app is preferable for an admin environment like this

**Reply by nickrum** [reply]
Oh, I didn't know that, I have never worked with v7. I actually worried about that as well while thinking about this idea. A smooth experience is definitely preferred. I guess for modules this would still work quite well, since they are more like different routes than various components on a single page loading randomly and jumping in.

Maybe, since you have been down that rabbit hole, it might be best to leave it as is 🐰

**Comment by rijkvanzanten** [comment]
This now exists!
